Added AutoNamedRouteAttribute

ZmorzynskiK opened the pull request:
Hi,

I noticed, that when there're many actions/controllers and when using MVC 4/5 attribute routing, there might be some performance issues.
If you want to know a little more, I wrote about it here:
http://stuffleftatparty.com/the-ugly-truth-behind-pretty-urls

That's why I decided to create some automatic mechanism to cope with this issue.
I added `AutoNamedRouteAttribute` that one can use instead of regular `RouteAttribute` on actions. This new attribute will automatically assign a name to a route so that it'll be added to MVC's internal dictionary.
To generate a link, instead of regular `Html.ActionLink`, one can use new `Html.AutoNamedActionLink` that will generate route link from name, resulting in much faster link generation.

Since T4MVCExtensions is dependent of MVC 3 and route attributes are present from MVC 4, therefore I added new test website based on MVC 5 that shows the idea.
The most important files are just 2 files in T4MVC folder: RouteExtensions.cs and AutoNamedRouteAttribute.cs

I'm wondering if anyone could find this stuff useful? And if yes, then how to incorporate it into T4MVC?

best,
Krzysztof

ZmorzynskiK commented:
Oh and sorry for the reverts, I accidentaly reverted whole commit, so then I had to revert that revert, so that's why there are that much commits :)

davidebbo commented:
I'd prefer not to add a second test project, as that'll cause maintenance pain, with having two separate copies of the .tt file.

MVC3 is ancient, so I'd have no problem moving it to a newer version if that helps keep a single project.

As an aside, make sure you use spaces and not tabs in all files as that's what the project is using.

ZmorzynskiK commented:
OK, so moving sample website to newest MVC version should be that much of a problem (I could even do it later), but since `AutoNamedRouteAttribute` uses feature available from MVC4 (I think), then we have the following options regarding T4MVCExtensions:
1. update MVC dependency to a newest version and implement `AutoNamedRouteAttribute` directly into T4MVCExtensions
2. create another project/assembly for MVC5, like T4MVCExtensions.MVC5 or something like that, and implement those changes there

Which way would you like to go?

As for spaces vs tabs - sure, I'll remember that.

best,
KZ

davidebbo commented:
I'm fine saying that T4MVCExtensions now takes a dependency on MVC4 instead of MVC3. I don't think it will affect many people at this point.

ZmorzynskiK commented:
I did a little more digging and it turned out, that attribute routing wasn't available until MVC 5.
So, should we update MVC dependency of T4MVCExtensions to MVC 5.0.0:
https://www.nuget.org/packages/Microsoft.AspNet.Mvc/5.0.0
? It is 3 years old already, so I presume it should be safe?

Or maybe just go with current MVC version (5.2.3)?

best,
KZ

ZmorzynskiK commented:
Hi David,

I'm not pushing or anything, but did you maybe managed to decide if we could update MVC dependency of T4MVCExtensions to MVC 5.0.0 or MVC 5.2.3 (current version)?

best regards,
KZ

davidebbo commented:
Yes, sorry, I think this is fine. We need to move forward, and generally don't have the resources to support a wide range of old versions. Users can always choose to install an older T4MVC.

ZmorzynskiK commented:
OK, so I'll soon update MVC dependency for T4MVCExtensions to version 5.0.0 and move some samples from old host website to the new one based on MVC 5.X.X and then delete the old MVC 3 website so we don't have  duplicated .tt files.

best,
KZ

davidebbo commented:
The key is to keep all the crazy views/files, which were added over the years to cover all kind of edge cases. One of the key ways used to prevent regression is to always carefully inspect the before/after diff when any changes are made to the .tt file, making sure that the diff matches expectations.

ZmorzynskiK commented:
Yes, I'll try to do it like you wrote: keep all the crazy views/files.
The .tt file won't change, so I'll just get the one that is currently in use in MVC3 host website.

ZmorzynskiK commented:
So I migrated `T4MVCExtensions` to MVC 5.2.3 (turned out that is was needed to update to this version), and added `AutoNamedRouteAttribute` and `Html.AutoNamedRouteLink` helper to it. I also added new sample MVC 5.2.3 website and tests for it. I then removed old MVC3 host website and old tests.

I copied all the files, crazy views etc from old MVC3 host website to new MVC5 host website and copied all the tests from old to new one. Plus I added "many controllers with `RouteAttribute`" sample to the website to compare link generation speeds.

I didn't touch `nuspec` file.

With all those changes, almost everything went smooth, except one test that now fails: `TestNoPublicClassesinGlobalNamespace` as you can see in CI report.
This is because I had to change access modifier in `T4MVC.tt.hooks.t4` file from:
`internal static class T4MVCHelpers`
to:
`public static class T4MVCHelpers`
because otherwise the tests project wouldn't compile.

Any idea what to do with that?

best,
KZ

davidebbo commented:
Since the new app is replacing the old one, can't its name (`T4MVCHostMvcApp`) be preserved? Right now the diff is enormous mostly because of the new name, and that makes it a little too scary for me :). But I would assume that the majority of files would remain identical in the new app, such that only the relevant changes would show up in the diff. Or can't the existing app simply be updated to use the new MVC?

davidebbo commented:
As for your question, the answer is [here](https://github.com/T4MVC/T4MVC/blob/7decd7d238415bb4dd86fe99bf6f036369d53907/T4MVCHostMvcApp/Properties/AssemblyInfo.cs#L6).

ZmorzynskiK commented:
OK, so I guess I should revert last commit (deleting old sample website and tests) and then and convert/merge "old" website to "new" website, but still under old name.

There are some changes (ie. new MVC5 template uses bootstrap etc), but yes, the majority should be the same as "old".
Or maybe just update to MVC5... maybe that's the way to go.

Ah, right, forgot about `InternalsVisibleTo`, thanks.

best,
KZ

ZmorzynskiK commented:
Hi,

as you suggested, I updated MVC3 sample website to use MVC5 + I added many controllers for speed testing with route attributes.
I also had to add new define switch (`USE_ROUTE_ATTRIBUTES`) since attribute routes mapping doesn't work with generic actions, so when this define is set (like it's right now) it excludes generic actions.

So the majority of the diff is just related to those many controllers that were added.
The tests didn't change (except for versions references in app.config).

best,
KZ

davidebbo commented:
Thanks, that's makes it a lot easier to look through. So there are no changes at all to the main t4mvc.tt file?

Question: if someone uses t4mvc today and upgrades to this, would you expect any diffs to their generated files (assuming no change in their code)? I'm guessing not if the .tt file didn't change.

I'm going to be out for a little while after today, so I won't be responsive. But let's work on getting that in after the new year.

Thanks for your work in this contribution!

@kevinkuszyk let us know is you have any opinion about this as well.

davidebbo commented:
Another question is whether we should move the T4MVC version to 4.0, since it's clearly a breaking change (since it won't work with projects using older MVC). Following semver, it feels like a clearcut case that we should.

ZmorzynskiK commented:
> Thanks, that's makes it a lot easier to look through. So there are no changes at all to the main t4mvc.tt file?
> Question: if someone uses t4mvc today and upgrades to this, would you expect any diffs to their generated files (assuming no change in their code)? I'm guessing not if the .tt file didn't change.

That's right, I didn't change T4MVC.tt file (or T4MVC.tt.hooks.t4, or T4MVC.tt.settings.xml) at all, therefore there shouldn't be any changes in generated files.

The only really meaningful changes where made to `T4MVCExtensions` project: added AutoNamedRouteAttribute.cs file and added some additional helpers to T4Extensions.cs file.

> Another question is whether we should move the T4MVC version to 4.0, since it's clearly a breaking change (since it won't work with projects using older MVC). Following semver, it feels like a clearcut case that we should.

That's entirely up to you :)

best,
KZ

kevinkuszyk commented:
@ZmorzynskiK thanks for the contribution.

@davidebbo looks good to me (admittedly I scanned through the changes - there's a lot there).

Regarding bumping the version number - yes, I agree that makes sense here.

@ZmorzynskiK is 5.2.3 the lowest version of MVC which supports these changes?  Where possible I always try to keep libraries open to the most number of consumers.  I think 5.0.0 would be better if we can support it.

I'm also going offline later this week for Christmas.  I'm happy to continue the conversation in the new year.

ZmorzynskiK commented:
I tried to use MVC 5.0.0 but unfortunately it turned out that it didn't have yet an `IDirectRouteFactory` interface that allows to create routes directly from action attribute. This interface (along with some more changes in routing) was added in 5.2.2 if I see correctly, but since there's a mess with ASP.NET versioning and source code repositories, there's a chance I might be wrong.

According to Nuget, the 5.2.2 was released 2014.08.28 (over 2 years ago), and 5.2.3 was released 2015.02.09 (**almost 2 years ago** - how time flies!) and it has the most downloads of all, so maybe it's safe to use 5.2.3 already?

And if someone is using older MVC then there's a good chance that him/her is not using route attributes at all and in this case, can stick with older T4MVCExtensions version (which is identical except this route attributes related changes).

Have a nice Holidays both of you.

best,
KZ

ZmorzynskiK commented:
Hi guys,

I was just wondering if you've had a moment to give those changes a thought and maybe accept this pull request if you're fine with it?

best,
KZ

davidebbo commented:
Sorry Krzysztof for the slow response! It may take a few more days before I get back to this, but it is not forgotten :)

davidebbo commented on T4MVCExtensions/AutoNamedRouteAttribute.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace System.Web.Mvc
+{
+    [AttributeUsage( AttributeTargets.Method | AttributeTargets.Class, Inherited = false, AllowMultiple = true )]
```
The project uses the default VS styling, which does not put spaces inside parenthesis. Can you reformat accordingly? Same in some other files.

ZmorzynskiK replied:
Please check if now is OK with formatting.

davidebbo commented:
Pushed to NuGet as 4.0.0-beta1. Can you give it a try and make sure everything works as it should? Thanks!

ZmorzynskiK commented:
Sure, I'll use it in a couple of my projects.
